src/qml/qml_owned_ptr.h

```diff
@@ -0,0 +1,155 @@
+// by the QML (or JavaScript) Engine, and thus such pointers do not require a
+// manual delete to free the heap memory when they go out of scope, as they will
+// be handled by the engine garbage collector.
+template<typename T>
+class qml_owned_ptr final {
+  public:
+    qml_owned_ptr() noexcept
+            : m_ptr{nullptr} {
+    }
+
+    qml_owned_ptr(std::nullptr_t) noexcept
+            : m_ptr{nullptr} {
+    }
+
+    explicit qml_owned_ptr(T* t) noexcept
+            : m_ptr{t} {
+        DEBUG_ASSERT(m_ptr);
+        QQmlEngine::setObjectOwnership(m_ptr, QQmlEngine::JavaScriptOwnership);
+    }
+
+    // explicitly generate trivial destructor (since decltype(m_ptr) is not a class type)
+    ~qml_owned_ptr() noexcept = default;
+
+    // Rule of 5
+    qml_owned_ptr(const qml_owned_ptr<T>&) = delete;
+    qml_owned_ptr& operator=(const qml_owned_ptr<T>&) = delete;
+    qml_owned_ptr(const qml_owned_ptr<T>&& other) = delete;
+    qml_owned_ptr& operator=(const qml_owned_ptr<T>&& other) = delete;
+
+    // If U* is convertible to T* then qml_owned_ptr<U> is convertible to qml_owned_ptr<T>
+    template<
+            typename U,
+            typename = typename std::enable_if<std::is_convertible<U*, T*>::value, U>::type>
+    qml_owned_ptr(qml_owned_ptr<U>&& u) noexcept
+            : m_ptr{u.m_ptr} {
+        u.m_ptr = nullptr;
+    }
+
+    // If U* is convertible to T* then qml_owned_ptr<U> is assignable to qml_owned_ptr<T>
+    template<
+            typename U,
+            typename = typename std::enable_if<std::is_convertible<U*, T*>::value, U>::type>
+    qml_owned_ptr& operator=(qml_owned_ptr<U>&& u) noexcept {
+        qml_owned_ptr temp{std::move(u)};
+        std::swap(temp.m_ptr, m_ptr);
+        DEBUG_ASSERT(!m_ptr ||
+                QQmlEngine::objectOwnership(m_ptr) ==
+                        QQmlEngine::JavaScriptOwnership);
+        return *this;
+    }
+
+    qml_owned_ptr& operator=(std::nullptr_t) noexcept {
+        qml_owned_ptr{std::move(*this)}; // move *this into a temporary that gets destructed
+        return *this;
+    }
+
+    // Prevent unintended invocation of delete on qml_owned_ptr
+    operator void*() const = delete;
+
+    operator T*() const noexcept {
+        return m_ptr;
+    }
+
+    T* get() const noexcept {
+        return m_ptr;
+    }
+
+    T& operator*() const noexcept {
+        return *m_ptr;
+    }
+
+    T* operator->() const noexcept {
+        return m_ptr;
+    }
+
+    operator bool() const noexcept {
+        return m_ptr != nullptr;
+    }
+
+    QPointer<T> toWeakRef() {
+        return m_ptr;
+    }
+
+  private:
+    T* m_ptr;
+
+    template<typename>
+    friend class qml_owned_ptr;
+};
+
+template<typename T, typename... Args>
+    requires(std::is_base_of_v<QObject, T>)
+inline qml_owned_ptr<T> make_qml_owned(Args&&... args) {
+    return qml_owned_ptr<T>(new T(std::forward<Args>(args)...));
+}
```

Comment:
```suggestion
template<typename T, typename... Args>
    requires(std::is_base_of_v<QObject, T>)
qml_owned_ptr<T> make_qml_owned(Args&&... args) {
    return qml_owned_ptr<T>(new T(std::forward<Args>(args)...));
}
```
👍 on the `requires`, `inline` is unnecessary though (all declarations in C++ headers are implicitly inline).